libs/langgraph/src/pregel/runner.ts

```diff
@@ -113,10 +133,137 @@ export class PregelRunner {
-      // placeholder function - will have logic added when functional API is implemented
-      return task.config?.configurable?.[CONFIG_KEY_SEND]?.(writes) ?? [];
+      if (writes.every(([channel]) => channel !== PUSH)) {
+        return task.config?.configurable?.[CONFIG_KEY_SEND]?.(writes) ?? [];
+      }
+
+      // TODO: type the scratchpad stuff, extract boilerplate
+      // Schedule PUSH tasks, collect promises
+      if (task.config?.configurable?.[CONFIG_KEY_SCRATCHPAD] == null) {
+        patchConfigurable(task.config, {
+          [CONFIG_KEY_SCRATCHPAD]: {},
+        });
+      }
+      const scratchpad = task.config?.configurable?.[
+        CONFIG_KEY_SCRATCHPAD
+      ] as Record<string, unknown>;
+      scratchpad.callCounter = (scratchpad.callCounter as number) ?? 0;
+
+      const rtn: Record<number, Promise<unknown> | undefined> = {};
+
+      for (const [idx, write] of writes.entries()) {
+        const [channel] = write;
+        if (channel !== PUSH) {
+          continue;
+        }
+
+        const wcall = calls?.[idx];
+        const cnt = scratchpad.callCounter as number;
+        scratchpad.callCounter = cnt + 1;
+
+        if (wcall == null) {
+          throw new Error("BUG: No call found");
+        }
+
+        const nextTask = this.loop.acceptPush(task, cnt, wcall);
+
+        if (!nextTask) {
+          continue;
+        }
+
+        // Check if this task is already running
+        const existingPromise = executingTasksMap[nextTask.id];
+
+        if (existingPromise !== undefined) {
+          // If the parent task was retried, the next task might already be running
+          rtn[idx] = existingPromise;
+        } else if (nextTask.writes.length > 0) {
+          // If it already ran, return the result
+          const returns = nextTask.writes.filter(([c]) => c === RETURN);
+          const errors = nextTask.writes.filter(([c]) => c === ERROR);
+
+          if (returns.length > 0) {
+            // Task completed successfully
+            if (returns.length === 1) {
+              rtn[idx] = Promise.resolve(returns[0][1]);
+            } else {
+              // the only way this should happen is if the task executes multiple times and writes aren't cleared
+              throw new Error(
+                `BUG: multiple returns found for task ${nextTask.name}__${nextTask.id}`
+              );
+            }
+          } else if (errors.length > 0) {
+            if (errors.length === 1) {
+              const errorValue = errors[0][1];
+              // Task failed
+              const error =
+                // eslint-disable-next-line no-instanceof/no-instanceof
+                errorValue instanceof Error
+                  ? errorValue
+                  : new Error(String(errorValue));
+
+              rtn[idx] = Promise.reject(error);
+            } else {
+              // the only way this should happen is if the task executes multiple times and writes aren't cleared
```

Comment:
or if it was executed without saving either error or return, but that's not expected either.